quickwit-indexing/src/models/indexing_sync.rs

```diff
@@ -0,0 +1,90 @@
+// Copyright (C) 2022 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::Arc;
+
+use tokio::sync::Mutex;
+
+/// Publishers must acquire this lock and ensure that the indexing generation is current before
+/// publishing.
+pub type PublishLock = Arc<Mutex<()>>;
+
+/// Indexing generation writer held by the source, when a partition reassignment occurs, the leader
+/// must:
+///
+/// 1. Acquire the publish lock.
+/// 2. Increment the indexing generation.
+/// 3. Release the publish lock.
+/// 4. Propagate to the downstream consumers the new current indexing generation object.
+#[derive(Debug, Default)]
+pub struct IndexingGenerationLeader(Arc<AtomicUsize>);
+
+impl IndexingGenerationLeader {
```

Comment:
Can we merge the lock and the generation stuff into one single object in order to enforce the contract described in the comment (both on the source / writer side and the actor side).